sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/Query13.java

```diff
@@ -0,0 +1,108 @@
+import java.io.IOException;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.nexmark.NexmarkConfiguration;
+import org.apache.beam.sdk.nexmark.model.Event;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.GroupByKey;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.Reshuffle.AssignShardFn;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+
+/**
+ * Query "13" PORTABILITY_BATCH (not in original suite).
+ *
+ * <p>This benchmark is created to stress the boundary of runner and SDK in portability world. The
+ * basic shape of this benchmark is source + GBK + ParDo, in which the GBK read + ParDo will require
+ * that runner reads from shuffle and connects with SDK to do CPU intensive computation.
+ */
+@SuppressWarnings({
+  "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+})
+public class Query13 extends NexmarkQueryTransform<Event> {
+  private final NexmarkConfiguration configuration;
+
+  public Query13(NexmarkConfiguration configuration) {
+    super("Query13");
+    this.configuration = configuration;
+  }
+
+  @Override
+  public PCollection<Event> expand(PCollection<Event> events) {
+    final Coder<Event> coder = events.getCoder();
+    return events
+        .apply("Pair with random key", ParDo.of(new AssignShardFn<>(configuration.numKeyBuckets)))
+        .apply(GroupByKey.create())
+        .apply(
+            "ExpandIterable",
+            ParDo.of(
+                new DoFn<KV<Integer, Iterable<Event>>, Event>() {
+                  @ProcessElement
+                  public void processElement(
+                      @Element KV<Integer, Iterable<Event>> element, OutputReceiver<Event> r) {
+                    for (Event value : element.getValue()) {
+                      r.output(value);
+                    }
+                  }
+                }))
+        // Force round trip through coder.
+        .apply(
+            name + ".Serialize",
+            ParDo.of(
+                new DoFn<Event, Event>() {
+                  private final Counter bytesMetric = Metrics.counter(name, "serde-bytes");
+                  private long elementCount = 0;
+                  private long opFrequency =
+                      (configuration.pardoCPUFactor > 0.0 && configuration.pardoCPUFactor < 1.0)
+                          ? (long) (1.0 / configuration.pardoCPUFactor)
+                          : 1L;
+
+                  @ProcessElement
+                  public void processElement(ProcessContext c) throws CoderException, IOException {
+                    elementCount++;
+                    Event event;
+                    if (configuration.pardoCPUFactor != 0
+                        && (opFrequency == 1L || elementCount % opFrequency == 0L)) {
```

Comment:
how about using Random.nextDouble()  compared with pardoCPUFactor?

Reply:
Do you mean always use randomization, instead of using a fixed pardoCPUFactor?

If so, I think have a way to control pardoCPUFactor is better at this moment For example, in near term usage, I would expect to always set `pardoCPUFactor` as 1.0 to make the ParDo do CPU intensive work

Reply:
I mean instead of using mod you can just achieve same effect with `if (Random.nextDouble() < configuration.pardoCPUFactor)` right? Plus this is more accurate because you don't have to round up the opFrequency.

Reply:
Thanks for the idea. I changed to use Randome.nextDouble() now